### PR TITLE
[NEP-18619]: Adjust the endpoint for Dashboard::Datasets::List to optionally include filter datasets

### PR DESCRIPTION
**Pull request by soundarya-mv**
Ticket: https://jobready.atlassian.net/browse/NEP-18619

This ticket modifies the `Superset::Dashboard::Datasets::List.new(id).dataset_details` and `Superset::Dashboard::Datasets::List.new(id).list` to optionally include filter datasets as well to duplicate those during the dashboard duplication process.

This also adds an additional check in source dataset duplication that if any filter dataset already exists with the new name in the target schema we can use the existing dataset itself for the new dashboard also.

### CODE REVIEW

File: lib/superset/dashboard/datasets/list.rb [hunk 1]
```diff
@@ -7,14 +7,15 @@ module Superset
...
       class List < Superset::Request
-        attr_reader :id  # dashboard id
+        attr_reader :id, :include_filter_datasets # dashboard id
 
         def self.call(id)
           self.new(id).list
         end
 
-        def initialize(id)
+        def initialize(id, include_filter_datasets = false)
```
Comment (jbat) [inline lib/superset/dashboard/datasets/list.rb:16]: @soundarya-mv 
as we are now passing more than one param, its probably a good idea to make these named attributes
```
  def initialize(dashboard_id: dashboard_id, include_filter_datasets: false)
```

File: lib/superset/dashboard/filters/list.rb [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+          @id = id
+        end
+
+        def perform
+          filters_configuration = JSON.parse(dashboard.result['json_metadata'])['native_filter_configuration'] || []
+          return Array.new unless filters_configuration && filters_configuration.any?
+  
+          # pull only the filters dataset ids from the dashboard
+          filters_configuration.map { |c| c['targets'] }.flatten.compact.map { |c| c['datasetId'] }.flatten.compact
+        end
```
Comment (jbat) [inline lib/superset/dashboard/filters/list.rb:17]: @soundarya-mv  I dont thing we  need a new class for this.

We can just add another method to the Dashboard::Get class

https://github.com/rdytech/superset-client/blob/develop/lib/superset/dashboard/get.rb#L25
```
      def json_metadata
        JSON.parse(result['json_metadata'])
      end

     def filter_configuration
         json_metadata['native_filter_configuration'] || []
     end

```

File: lib/superset/dashboard/datasets/list.rb [hunk 3]
```diff
@@ -36,13 +37,34 @@ def schemas
...
+          chart_dataset_ids = chart_datasets.map{|d| d['id'] }
+          filter_dataset_ids_not_used_in_charts = filter_dataset_ids - chart_dataset_ids
+          return chart_datasets if filter_dataset_ids_not_used_in_charts.empty?
+          dashboard_datasets = chart_datasets + filter_datasets(filter_dataset_ids_not_used_in_charts)
         end
 
         private
 
+        def filter_dataset_ids
+          @filter_dataset_ids ||= Dashboard::Filters::List.new(id).perform
```
Comment (jbat) [inline lib/superset/dashboard/datasets/list.rb:53]: See comment below .. we could memoize a dashboard object
and use it for the `dashboard.title` method as well as `dashboard.filter_configuration`

Just trying to minimise the number of API calls we use

      def dashboard
        @dashboard ||= Superset::Dashboard::Get.new(dashboard_id)
      end

File: lib/superset/dashboard/datasets/list.rb [hunk 4]
```diff
@@ -36,13 +37,34 @@ def schemas
...
         def datasets_details
-          result.map do |details|
+          chart_datasets = result.map do |details|
             details.slice('id', 'datasource_name', 'schema', 'sql').merge('database' => details['database'].slice('id', 'name', 'backend')).with_indifferent_access
           end
+          return chart_datasets unless include_filter_datasets
+          chart_dataset_ids = chart_datasets.map{|d| d['id'] }
+          filter_dataset_ids_not_used_in_charts = filter_dataset_ids - chart_dataset_ids
+          return chart_datasets if filter_dataset_ids_not_used_in_charts.empty?
+          dashboard_datasets = chart_datasets + filter_datasets(filter_dataset_ids_not_used_in_charts)
```
Comment (jbat) [inline lib/superset/dashboard/datasets/list.rb:47]: @soundarya-mv 
no real need to have another variable here, its not used any where
I would say use a comment if you want to be obvious that we are returning chart + filter datasets.

```
   ...
   return chart_datasets if filter_dataset_ids_not_used_in_charts.empty?
   
   # returning chart and filter datasets
   chart_datasets + filter_datasets(filter_dataset_ids_not_used_in_charts)
end

Comment (soundarya-mv) [reply lib/superset/dashboard/datasets/list.rb:47]: I also thought of removing it. Looks like I missed it. Will update it.

Comment (soundarya-mv) [reply lib/superset/dashboard/datasets/list.rb]: Have updated it now

File: lib/superset/dashboard/datasets/list.rb [hunk 5]
```diff
@@ -66,7 +89,11 @@ def rows
...
         def title
-          @title ||= [id, Superset::Dashboard::Get.new(id).title].join(' ')
+          @title ||= [id, dashboard.title].join(' ')
+        end
+
+        def dashboard
+          @dashboard = Superset::Dashboard::Get.new(id)
         end
       end
     end
```
Comment (jbat) [inline lib/superset/dashboard/datasets/list.rb:99]: Ideally the list method should also show filter only datasets
```
Superset::Dashboard::Datasets::List.new(dashboard_id: 21,  include_filter_datasets: true ).list
```

**Review by jbat — APPROVED** [review]

File: lib/superset/dashboard/bulk_delete_cascade.rb [hunk 6]
```diff
@@ -31,7 +31,7 @@ def perform
       private
 
       def delete_datasets(dashboard_id)
-        datasets_to_delete = Superset::Dashboard::Datasets::List.new(dashboard_id).datasets_details.map{|d| d[:id] }
+        datasets_to_delete = Superset::Dashboard::Datasets::List.new(dashboard_id: dashboard_id).datasets_details.map{|d| d[:id] }
         Superset::Dataset::BulkDelete.new(dataset_ids: datasets_to_delete).perform if datasets_to_delete.any?
```
Comment (jbat) [inline lib/superset/dashboard/bulk_delete_cascade.rb:35]: @soundarya-mv 

flagging that we will need another Ticket to not allow a BulkDelete 
if a chart or dataset is used on a different dashboard.

as currently there are no checks around this

File: lib/superset/services/duplicate_dashboard.rb [hunk 7]
```diff
@@ -259,7 +259,7 @@ def source_dashboard_dataset_ids
       end
 
       def source_dashboard_filter_dataset_ids
-        @filter_dataset_ids ||= Superset::Dashboard::Filters::List.new(id).perform
+        @filter_dataset_ids ||= source_dashboard.filter_configuration.map { |c| c['targets'] }.flatten.compact.map { |c| c['datasetId'] }.flatten.compact
```
Comment (jbat) [inline lib/superset/services/duplicate_dashboard.rb:262]: @soundarya-mv 

Just wondering if we should do a uniq here?
incase the same dataset is used across multiple filters

Comment (soundarya-mv) [reply lib/superset/services/duplicate_dashboard.rb]: Done

**Review by jbat — APPROVED** [review]

**Pull request merged**